core/trino-main/src/main/java/io/trino/exchange/ExchangeManagerRegistry.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.exchange;
+
+import io.airlift.log.Logger;
+import io.trino.metadata.HandleResolver;
+import io.trino.spi.TrinoException;
+import io.trino.spi.classloader.ThreadContextClassLoader;
+import io.trino.spi.exchange.ExchangeManager;
+import io.trino.spi.exchange.ExchangeManagerFactory;
+
+import javax.inject.Inject;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static io.airlift.configuration.ConfigurationLoader.loadPropertiesFrom;
+import static io.trino.spi.StandardErrorCode.EXCHANGE_MANAGER_NOT_CONFIGURED;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class ExchangeManagerRegistry
+{
+    private static final Logger log = Logger.get(ExchangeManagerRegistry.class);
+
+    private static final File CONFIG_FILE = new File("etc/exchange-manager.properties");
+    private static final String EXCHANGE_MANAGER_NAME_PROPERTY = "exchange-manager.name";
+
+    private final HandleResolver handleResolver;
+
+    private final Map<String, ExchangeManagerFactory> exchangeManagerFactories = new ConcurrentHashMap<>();
+
+    private volatile ExchangeManager exchangeManager;
```

Comment:
add `@GuardedBy("this")`

Reply:
In the `getExchangeManager` it is not guarded by the lock so the annotation might be misleading.